sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGenerateExec.scala

```diff
@@ -604,46 +602,49 @@ abstract class GpuExplodeBase extends GpuUnevaluableUnaryExpression with GpuGene
         indexIntoData = 0
         closeCurrent()
         if (inputIterator.hasNext) {
-          currentBatch = inputIterator.next()
+          spillableCurrentBatch = SpillableColumnarBatch(inputIterator.next(),
+            SpillPriorities.ACTIVE_BATCHING_PRIORITY)
         }
       }
 
       override def hasNext: Boolean = {
-        if (currentBatch == null || indexIntoData >= numArrayColumns) {
+        if (spillableCurrentBatch == null || indexIntoData >= numArrayColumns) {
           fetchNextBatch()
         }
-        currentBatch != null
+        spillableCurrentBatch != null
       }
 
       override def next(): ColumnarBatch = {
-        if (currentBatch == null || indexIntoData >= numArrayColumns) {
-          fetchNextBatch()
-        }
-
         withResource(new NvtxWithMetrics("GpuGenerateExec", NvtxColor.PURPLE, opTime)) { _ =>
-          withResource(new Array[ColumnVector](numExplodeColumns + numOtherColumns)) { result =>
-
-            withResource(GpuProjectExec.project(currentBatch, boundOthersProjectList)) { cb =>
-              (0 until cb.numCols()).foreach { i =>
-                result(i) = cb.column(i).asInstanceOf[GpuColumnVector].getBase.incRefCount()
-              }
-            }
-            if (position) {
-              result(numOtherColumns) = withResource(GpuScalar.from(indexIntoData, IntegerType)) {
-                scalar => ColumnVector.fromScalar(scalar, currentBatch.numRows())
+          // Do projection with boundOthersProjectList and current boundLazyProjectList item
+          val projectCb = GpuProjectExec.projectAndCloseWithRetrySingleBatch(
+            SpillableColumnarBatch(
+              spillableCurrentBatch.getColumnarBatch(),
+              SpillPriorities.ACTIVE_ON_DECK_PRIORITY),
+            boundOthersProjectList :+ boundLazyProjectList(indexIntoData))
+          withResource(projectCb) { _ =>
+            withResource(new Array[ColumnVector](numOtherColumns + numExplodeColumns)) { result =>
+              val projCols = GpuColumnVector.extractColumns(projectCb)
+              // Copy columns corresponding to boundOthersProjectList
+              (0 until numOtherColumns).foreach { i =>
+                result(i) = projCols(i).getBase.incRefCount()
               }
-            }
-            result(numOtherColumns + numExplodeColumns - 1) =
-                withResource(GpuProjectExec.project(currentBatch,
-                  Seq(boundLazyProjectList(indexIntoData)))) { cb =>
-                  cb.column(0).asInstanceOf[GpuColumnVector].getBase.incRefCount()
+              // Add position column, if needed.
+              if (position) {
+                result(numOtherColumns) = withResource(GpuScalar.from(indexIntoData, IntegerType)) {
+                  scalar => ColumnVector.fromScalar(scalar, projectCb.numRows())
```

Comment:
nit: This allocates memory. Do we want to have a retry around this too??? It feels like something we could probably put into the projection that we do initially, with the retry.

Reply:
Thank you for this hint @revans2!  This makes the code much simpler.